site/content/features/records.md

```diff
@@ -0,0 +1,366 @@
+      System.out.println("Fields: " + java.util.Arrays.toString(fields));
+      RecordComponent[] components = Point.class.getRecordComponents();
+      System.out.println("Components: " + java.util.Arrays.toString(components));
+
+      var p = new Point(3, 4);
+      RecordComponent rc = p.getClass().getRecordComponents()[0];
+      Object value = rc.getAccessor().invoke(p);
+      System.out.println("Component " + rc.getName() + " has value " + value);
+   }
+}
+{{< /sandboxsource >}}
+{{< /sandbox >}}
+
+Some Further Observations
+-------------------------
+
+1\.  Some languages have tuples or product types. In those languages, you can model a point as a pair of `double`. But in Java, we like names. Point components should have names `x` and `y`, and we want the whole thing to be a `Point`, distinct from any other pairs of `double`.
+
+2\. A record variable holds a reference to an object. That is, records are _not_ value or inline types—another new kid on the block. Project Valhalla will let you define
+
+```
+inline class Point {
+   private double x;
+   private double y;
+   ...
+}
+```
+
+Then a `Point` variable holds a flat 16 bytes of data, not a reference to an object. But the fields are still encapsulated. In time, you should be able to declare an `inline record`, with flat layout and no encapsulation.
+
+3\. Records are only as immutable as their fields are. Nothing stops you from having mutable components:
+
+{{< sandbox version="java14" preview="true" mainclass="RecordDemo" >}}
+{{< sandboxsource "RecordDemo.java" >}}
+record Point(double... coordinates) {
+   public String toString() {
+      return java.util.Arrays.toString(coordinates);
+   }
+}
+
+public class RecordDemo {
+   public static void main(String[] args) {
+      Point p = new Point(0, 0);
+      System.out.println("The origin: " + p);
+      p.coordinates()[0] = 3;
+      System.out.println("Not anymore: " + p);
+   }
+}
+{{< /sandboxsource >}}
+{{< /sandbox >}}
+
+Because arrays are mutable, you can change the elements of `coordinates`.
+
+This is not a good idea, but the Java language won't stop you. In general, Java has no mechanism for expressing immutability. It is no different with reccords.
+
+4\. The implementations of `hashCode`, `equals`, and `toString` in the JDK are _not normative_. In particular, the current behavior of combining two hash codes as `31 * h1 + h2` could change. The behavior of `equals` is constrained by the general `Object.equals` contract, but there is no guarantee that the order of comparisons is fixed. You should not rely on the exact format of the`toString` result either.
+
+5\. It is envisioned that in the future, records can be used for pattern matching, with a syntax somewhat like:
+
+```
+switch (obj) { 
+   case instanceof Point(x, 0) p: ... // Maybe the future—not in JDK 14
+   ...
+}
+```
+
+
+
+## References
+
+* [JEP 395: Records (Preview), OpenJDK](http://openjdk.java.net/jeps/395)
```

Comment:
Link text and link itself should probably be `359`

Reply:
Yes